projects/components/src/table/table.component.ts

```diff
@@ -409,8 +407,8 @@ export class TableComponent
   }
 
   private buildDataSource(): TableCdkDataSource | undefined {
-    if (!this.data || !this.columnConfigs) {
-      throw new Error('Undefined data or columnConfigs');
+    if (!this.canBuildDataSource()) {
```

Comment:
I'd say I feel like 6/10. I'm convinced, but your call. I think there's a difference from the usual debate here (or maybe phrased, we have a new lens on the usual). We know that there are valid states (e.g. async inputs) that can lead to a data source being unbuildable at any point in time. When those states happen, we clear any current data source. So what we've done is separated the logic into multiple checks rather than a single check - we first check if it's ok to build the data source, then branch, to either unset it or build it. That branch is duplicated inside building the data source to either error or build it. So I'm arguing to collapse to a single branch so the two aren't required to match and there is no error condition.

Trying to phrase it slightly differently, this isn't an error triggered by how the component is used, but an error internal to the component's implementation that we could tweak to make no error situation possible.